src/store/token/actions.js

```diff
@@ -203,29 +204,86 @@ export function requestUBIPayout(payout) {
   };
 }
 
-async function loopTransfer(from, to, value, paymentNote) {
-  return await waitAndRetryOnFail(
-    () => {
-      return core.token.transfer(from, to, value, paymentNote);
-    },
-    () => {
-      return true;
-    },
-    {},
-    () => {
-      return core.token.updateTransferSteps(from, to, value);
-    },
-  );
+// Recursive helper function for transfer
+async function loopTransfer(
+  from,
+  to,
+  value,
+  paymentNote,
+  hops,
+  attemptsLeft,
+  errorsMessages = '',
+) {
+  if (attemptsLeft === 0 || hops === 0) {
+    // ran out of attempts or hops, cannot attempt further transfers
+    throw new TransferError(errorsMessages);
+  }
+  try {
+    return await core.token.transfer(from, to, value, paymentNote, hops);
+  } catch (error) {
+    // RETRY when path is not found or is too long
+    // UPDATE edges database when the path fails and then retry
+    // GIVE UP for other errors
+    // ---
+    // no path or path too long
+    if (
+      error.name === 'TransferError' &&
+      (error.code === ErrorCodes.TOO_COMPLEX_TRANSFER || // too many steps in found path
+        error.code === ErrorCodes.UNKNOWN_ERROR || // includes timeout error from api
+        error.code === ErrorCodes.INVALID_TRANSFER) // other errors from find transitive transfer
+    ) {
+      // retry with fewer hops
+      return await loopTransfer(
```

Comment:
maybe would be worth to update the edges of db here as well before looping again?

Reply:
no because it will take more time, and there is no indication that edges are outdated, just that it is either too many hops or takes to long to calculate. most likely update edges will also time out due to the same reason. We use update when a found path is not accepted as transfer

Reply:
https://github.com/CirclesUBI/circles-myxogastria/blob/29599c9ae2e28ac7e1c0aaa961f641cc858a769e/src/views/SendConfirm.js#L137